### PR TITLE
UI/UX for hiding input/output

### LINKED ISSUE #713: UI/UX for collapsing input/output
This is an issue to track the UI/UX aspects of hiding showing some or all of a cell (input, output or both).

### PR DESCRIPTION
**Pull request by ellisonbg**
Fixes #713

### CODE REVIEW

**Comment by ellisonbg** [comment]
OK, this is ready for review. I think we should land this as quickly as possible so folks can play with the behavior and @cameronoelsen can follow on with some design work. I will open up follow on issues after this is merged.

Here are some of the UX questions we will have to answer in subsequent PRs:

* Should a cell's input be shown if it is run? Right now, it isn't. My thinking was to allow a user to run a notebook without showing the code.
*...

**Comment by ellisonbg** [comment]
One other thing we will need to follow up with. Right now if the `input|outputHidden` state on a cell changes from the actions (not a user's click on the collapser), the collapser won't re-render. Need to wire up a signal that the collapser can watch to trigger that.

File: packages/cells/src/collapser.tsx [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+/**
+ * A collapser subclass to collapse a cell's input area.
+ */
+export
+class InputCollapser extends Collapser {
+  /**
+   * Construct a new input collapser.
+   */
+  constructor() {
+      super();
```
Comment (blink1073) [inline packages/cells/src/collapser.tsx]: indent is too big

File: packages/cells/src/collapser.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+
+}
+
+/**
+ * A collapser subclass to collapse a cell's output area.
+ */
+export
+class OutputCollapser extends Collapser {
+    /**
+    * Construct a new output collapser.
```
Comment (blink1073) [inline packages/cells/src/collapser.tsx]: should be indented one more space

File: packages/cells/src/headerfooter.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * The interface for a cell footer.
+ */
+export
+interface ICellFooter extends Widget {}
+
+
+/**
+ * Default implementation of a cell header.
```
Comment (blink1073) [inline packages/cells/src/headerfooter.ts]: cell footer*

File: packages/cells/src/placeholder.tsx [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+  h, VirtualNode
+} from '@phosphor/virtualdom';
+
+import {
+  VDomRenderer
+} from '@jupyterlab/apputils';
+
+
+/**
+ * The CSS class added to placehodlers.
```
Comment (blink1073) [inline packages/cells/src/placeholder.tsx]: placeholders*

File: packages/cells/src/widget.ts [hunk 5]
```diff
@@ -253,13 +269,37 @@ class Cell extends Widget {
...
   }
-  set inputCollapsed(value: boolean) {
-    this._inputCollapsed = value;
+  set inputHidden(value: boolean) {
+    if (this._inputHidden === value) {
+      return;
+    }
+    let layout = this._inputWrapper.layout as PanelLayout;
+    if (value) {
+      layout.removeWidget(this._input);
```
Comment (blink1073) [inline packages/cells/src/widget.ts]: We should not be calling `removeWidget` directly, per the [docs](https://github.com/phosphorjs/phosphor/blob/06f24b7429e79da108e909f4aa1ba40661fe9372/packages/widgets/src/layout.ts#L175).  Can this be handled with show/hide?

File: packages/cells/style/headerfooter.css [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+| Distributed under the terms of the Modified BSD License.
+|----------------------------------------------------------------------------*/
+
+
+/*-----------------------------------------------------------------------------
+| Header/Footer
+|----------------------------------------------------------------------------*/
+
+
+/* Hidden by zero height be default */
```
Comment (blink1073) [inline packages/cells/style/headerfooter.css]: by default (sp)?

File: packages/cells/src/collapser.tsx (hunk 1 above)
Comment (ellisonbg) [inline packages/cells/src/collapser.tsx]: fixed...

File: packages/cells/src/collapser.tsx (hunk 2 above)
Comment (ellisonbg) [inline packages/cells/src/collapser.tsx]: fixed

File: packages/cells/src/headerfooter.ts (hunk 3 above)
Comment (ellisonbg) [inline packages/cells/src/headerfooter.ts]: fixed

File: packages/cells/src/placeholder.tsx (hunk 4 above)
Comment (ellisonbg) [inline packages/cells/src/placeholder.tsx]: fixed

File: packages/cells/src/widget.ts (hunk 5 above)
Comment (ellisonbg) [inline packages/cells/src/widget.ts]: I think it is important in this case to not have the elements in the DOM, switching to setting `parent` to `null`.

File: packages/cells/style/headerfooter.css (hunk 6 above)
Comment (ellisonbg) [inline packages/cells/style/headerfooter.css]: fixed

**Comment by ellisonbg** [comment]
All review comments addressed.

**Comment by ellisonbg** [comment]
Should be ready.

**Comment by blink1073** [comment]
Great, thanks!

**Pull request merged**